### PR TITLE
[1LP][RFR] Add services loading time performance test.

### PR DESCRIPTION
**Pull request by jaryn**
Adding test. This will test
Bug 1688937 - proxy timeout loading list of services post update

This needs #8749

{{ py.test: -v  cfme/tests/services/test_service_performance.py cfme/tests/optimize/test_bottlenecks.py cfme/tests/candu/test_graph_groupbytag.py   }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 8ebc6280800d30a18a3f03586ad4d1afb152feb8:

`cfme/tests/test_performance.py`:
- :red_circle: Line 4:1: **F401** *'wait_for.wait_for' imported but unused*
- :red_circle: Line 20:16: **F821** *undefined name 'self'*
- :red_circle: Line 20:32: **F821** *undefined name 'self'*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit fa23a6b03fa67235b87909233d9d3af479dc5b81:

`cfme/tests/test_db_migrate.py`:
- :bangbang: Line 119:1: **E302** *expected 2 blank lines, found 1*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by mshriver** [comment]
yaml MR has been merged.

File: cfme/tests/test_performance.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import pytest
+
+from cfme.base.ui import navigate_to
+from cfme.services.myservice import MyService
+from cfme.tests.test_db_migrate import download_and_migrate_db
+from cfme.utils.conf import cfme_data
+
+
+class TimedCall(object):
```
Comment (mshriver) [inline cfme/tests/test_performance.py]: Did you consider using `timeit` here?

Comment (jaryn) [reply cfme/tests/test_performance.py]: Sure, but it accepts only strings of python code. So it is cumbersome to use here.

```
In [1]: o = object()

In [2]: import timeit

In [3]: timeit.timeit("str(o)")
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-d7f20afbb7aa> in <module>()
----> 1 timeit.timeit("str(o)")

/usr/lib64/python2.7/timeit.pyc in timeit(stmt, setup, timer, number)
    235            number=default_number):
    236     """Convenience function to create Timer object and call timeit method."""
--> 237     return Timer(stmt, setup, timer).timeit(number)
    238 
    239 def repeat(stmt="pass", setup="pass", timer=default_timer,

/usr/lib64/python2.7/timeit.pyc in timeit(self, number)
    200         gc.disable()
    201         try:
--> 202             timing = self.inner(it, self.timer)
    203         finally:
    204             if gcold:

/usr/lib64/python2.7/timeit.pyc in inner(_it, _timer)

NameError: global name 'o' is not defined
```

Comment (jaryn) [reply cfme/tests/test_performance.py]: Well I tried and found it is undocumented feature, that it accepts also callable. I wonder is it good idea to use it.

**Comment by jaryn** [comment]
@mshriver Hi. I moved the new test under the perf dir, but this means this test will not run unless the `--perf` argument for py.test is used, but it is not the case for example for this run:
https://cfmeqe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/downstream-510z-tests-master/222/consoleFull

So the question is: when are the perf tests ran? Should this test really be under the perf dir given that it seems it won't be normally run ?

File: cfme/base/ui.py [hunk 2]
```diff
@@ -339,17 +341,19 @@ def prerequisite(self):
...
         # Can be either blank or logged in
-        from cfme.utils import browser
         logged_in_view = self.create_view(BaseLoggedInPage)
         if logged_in_view.logged_in:
             logged_in_view.logout()
-        if not self.view.is_displayed:
+        # The login screen may not be completely loaded here, but it loads
+        # with time without needing a refresh.
+        try:
+            wait_for(lambda: self.view.is_displayed, num_sec=60, delay=1)
```
Comment (digitronik) [inline cfme/base/ui.py]: Please try like: as now view have `wait_displayed` method
```
try:
    view.wait_displayed('60s')
except TimedOutError:
   pass
```

File: cfme/utils/ssh.py [hunk 3]
```diff
@@ -380,11 +379,6 @@ def write_output(line, file):
             if exit_status != 0:
                 logger.warning('Exit code %d!', exit_status)
             return SSHResult(rc=exit_status, output=''.join(output), command=command)
-        except paramiko.SSHException:
```
Comment (digitronik) [inline cfme/utils/ssh.py]: Q: why we are removing `paramiko` exception? I think better to handle both as the session comes from paramiko.

File: cfme/utils/ssh.py [hunk 4]
```diff
@@ -262,14 +262,13 @@ def get_transport(self, *args, **kwargs):
...
 
-    def run_command(self, command, timeout=RUNCMD_TIMEOUT, reraise=False, ensure_host=False,
+    def run_command(self, command, timeout=RUNCMD_TIMEOUT, ensure_host=False,
                     ensure_user=False, container=None):
         """Run a command over SSH.
 
         Args:
             command: The command. Supports taking dicts as version picking.
             timeout: Timeout after which the command execution fails.
-            reraise: Does not muffle the paramiko exceptions in the log.
```
Comment (digitronik) [inline cfme/utils/ssh.py]: Q: can you please explain why we are removing `reraise`?

Comment (jaryn) [reply cfme/utils/ssh.py]: sorry for this inconvenience but please check https://github.com/ManageIQ/integration_tests/pull/8749. I had to create separate PR for this commit and the reason is written in the PR desc.

File: cfme/utils/ssh.py (hunk 3 above)
Comment (jaryn) [reply cfme/utils/ssh.py]: same as above, but in short... when `paramiko.SSHException` is raised, we will try to reconnect rather then just cover the exception under the carpet and then return the misleading `SSHResult` with `rc=1`. For example: When SSH failed to connect, we didn't run any command, thus returning `SSHResult` with rc is just wrong.

**Comment by jaryn** [comment]
This PR needs #8777

**Comment by dajoRH** [comment]
Lint report for commit 860da8535a3188a70797fff3765c9eab9287b0cd:

Lint OK! :cake: :punch: :cookie: .

* 860da8535a3188a70797fff3765c9eab9287b0cd
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which we do not like in PRs so please, get rid of it.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit dd60a2b2eee1018f68ba4ddfdc6e22dcebb465b2:

Lint OK! :cake: :punch: :cookie: .

* dd60a2b2eee1018f68ba4ddfdc6e22dcebb465b2
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which we do not like in PRs so please, get rid of it.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a6d8df84609646bb1832123fe276572bb8bbb474:

Lint OK! :cake: :punch: :cookie: .

* a6d8df84609646bb1832123fe276572bb8bbb474
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which we do not like in PRs so please, get rid of it.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 312340f391c7bc1aab85cebf360f6640203ea705:

Lint OK! :cake: :punch: :cookie: .

* 312340f391c7bc1aab85cebf360f6640203ea705
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which we do not like in PRs so please, get rid of it.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 54acb44fd509e518d223252493875480234d527e:

Lint OK! :cake: :punch: :cookie: .

* 54acb44fd509e518d223252493875480234d527e
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which we do not like in PRs so please, get rid of it.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4703c498b2752855e4a9aef47952a0cab0ced946:

Lint OK! :cake: :punch: :cookie: .

* 4703c498b2752855e4a9aef47952a0cab0ced946
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which we do not like in PRs so please, get rid of it.

*CFME QE Bot*

File: cfme/tests/perf/test_performance.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+@pytest.fixture
+def appliance_with_performance_db(temp_appliance_extended_db):
+    app = temp_appliance_extended_db
+    db_backups = cfme_data['db_backups']
+    performance_db = db_backups['performance_510']
+    download_and_migrate_db(app, performance_db.url, performance_db.desc)
+    yield app
+
+
+def test_services_performance(appliance_with_performance_db):
```
Comment (digitronik) [inline cfme/tests/perf/test_performance.py]: need `test_requirements` marker

Comment (jaryn) [reply cfme/tests/perf/test_performance.py]: @mshriver @digitronik I don't really know what requirement should I link this test with.

Comment (digitronik) [reply cfme/tests/perf/test_performance.py]: @jaryn I would link this test to specific Focus Area requirement like if this test belongs to `Service FA` then link to `service`.
```
@test_requirements.service
def test_services_performance(appliance_with_performance_db):
```

If we don't have test_requinment for FA; please feel free to create a new one.

**Comment by mshriver** [comment]
@jaryn considering location and running of the test, a separate team is responsible for the `perf` directory tests, and their running. They run on instances of MIQ/CFME that have specific compute resources in order to standardize performance results.

I think given the scope of this test, and attachment to BZ concerning this load time, that we can locate this module in `cfme/tests/services` and include it in our regular runs without running with all performance tests.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 8e6c87cbd4988102c77cd75e3499c09d021d677d

The *global* fixture **`configure_fleecing_modscope`** is used in the following files:
- **cfme/tests/control/test_compliance.py**
    - *compliance_vm*

The *global* fixture **`temp_appliance_extended_db`** is used in the following files:
- **cfme/tests/candu/test_graph_groupbytag.py**
    - *host*
- **cfme/tests/optimize/test_bottlenecks.py**
    - *db_tbl*
    - *db_events*
    - *db_restore*
    - *test_bottlenecks_report_event_groups*
    - *test_bottlenecks_report_show_host_events*
    - *test_bottlenecks_report_time_zone*
    - *test_bottlenecks_summary_event_groups*
    - *test_bottlenecks_summary_show_host_events*
    - *test_bottlenecks_summary_time_zone*
- **cfme/tests/services/test_service_performance.py**
    - *appliance_with_performance_db*

The *local* fixture **`appliance_with_performance_db`** is used in the following files:
- **cfme/tests/services/test_service_performance.py**
    - *test_services_performance*

Please, consider creating a PRT run against these tests make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
Lint report for commit 4e2022e36a7732bcb60e32656f5f70b4bdf0a171:

`cfme/base/ui.py`:
- :red_circle: Line 1781:18: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 1782:9: **F821** *undefined name 'Version'*
- :red_circle: Line 1782:27: **F821** *undefined name 'Image'*
- :red_circle: Line 1785:17: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 1786:9: **F821** *undefined name 'Version'*
- :red_circle: Line 1786:27: **F821** *undefined name 'Image'*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/services/test_service_performance.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+from cfme.base.ui import navigate_to
+from cfme.services.myservice import MyService
+from cfme.tests.test_db_migrate import download_and_migrate_db
+from cfme.utils.conf import cfme_data
+
+
+@pytest.fixture
+def appliance_with_performance_db(temp_appliance_extended_db):
+    app = temp_appliance_extended_db
+    db_backups = cfme_data['db_backups']
```
Comment (mshriver) [inline cfme/tests/services/test_service_performance.py]: This lookup needs to be done in a way that doesn't produce a KeyError when the field isn't available in cfme_data.

File: cfme/tests/services/test_service_performance.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    performance_db = db_backups['performance_510']
```
Comment (mshriver) [inline cfme/tests/services/test_service_performance.py]: Should this key be parametrized? Does it depend on the appliance version under test?

**Comment by dajoRH** [comment]
I detected some fixture changes in commit d4cd26400033f6df1cdd8e0142fabc0fbf575e35

The *global* fixture **`configure_fleecing_modscope`** is used in the following files:
- **cfme/tests/control/test_compliance.py**
    - *compliance_vm*

The *global* fixture **`temp_appliance_extended_db`** is used in the following files:
- **cfme/tests/candu/test_graph_groupbytag.py**
    - *host*
- **cfme/tests/optimize/test_bottlenecks.py**
    - *db_tbl*
    - *db_events*
    - *db_restore*
    - *test_bottlenecks_report_event_groups*
    - *test_bottlenecks_report_show_host_events*
    - *test_bottlenecks_report_time_zone*
    - *test_bottlenecks_summary_event_groups*
    - *test_bottlenecks_summary_show_host_events*
    - *test_bottlenecks_summary_time_zone*
- **cfme/tests/services/test_service_performance.py**
    - *appliance_with_performance_db*

The *local* fixture **`appliance_with_performance_db`** is used in the following files:
- **cfme/tests/services/test_service_performance.py**
    - *test_services_performance*

Please, consider creating a PRT run against these tests make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 2ae0ad370943c4572e5c403e9d3307b9ddacb648

The *global* fixture **`temp_appliance_extended_db`** is used in the following files:
- **cfme/tests/candu/test_graph_groupbytag.py**
    - *host*
- **cfme/tests/optimize/test_bottlenecks.py**
    - *db_tbl*
    - *db_events*
    - *db_restore*
    - *test_bottlenecks_report_event_groups*
    - *test_bottlenecks_report_show_host_events*
    - *test_bottlenecks_report_time_zone*
    - *test_bottlenecks_summary_event_groups*
    - *test_bottlenecks_summary_show_host_events*
    - *test_bottlenecks_summary_time_zone*
- **cfme/tests/services/test_service_performance.py**
    - *appliance_with_performance_db*

The *local* fixture **`appliance_with_performance_db`** is used in the following files:
- **cfme/tests/services/test_service_performance.py**
    - *test_services_performance*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/services/test_service_performance.py (hunk 6 above)
Comment (jaryn) [reply cfme/tests/services/test_service_performance.py]: done

File: cfme/tests/services/test_service_performance.py (hunk 7 above)
Comment (jaryn) [reply cfme/tests/services/test_service_performance.py]: no. the DB with many services pre-created is migrated to the required version

**Comment by dajoRH** [comment]
I detected some fixture changes in commit ae81ac922333f4a4328dd3ac9f0a3e2a7cb7ac97

The *global* fixture **`temp_appliance_extended_db`** is used in the following files:
- **cfme/tests/candu/test_graph_groupbytag.py**
    - *host*
- **cfme/tests/optimize/test_bottlenecks.py**
    - *db_tbl*
    - *db_events*
    - *db_restore*
    - *test_bottlenecks_report_event_groups*
    - *test_bottlenecks_report_show_host_events*
    - *test_bottlenecks_report_time_zone*
    - *test_bottlenecks_summary_event_groups*
    - *test_bottlenecks_summary_show_host_events*
    - *test_bottlenecks_summary_time_zone*
- **cfme/tests/services/test_service_performance.py**
    - *appliance_with_performance_db*

The *local* fixture **`appliance_with_performance_db`** is used in the following files:
- **cfme/tests/services/test_service_performance.py**
    - *test_services_performance*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by mshriver** [comment]
@jaryn sorry I missed your question about perf directory earlier. There are some automatic markers applied causing the collection you're talking about. I think its best you don't place your test in the perf folder, as these are specific workload tests that are built to monitor and report on memory usage by the workers during those workloads.

File: cfme/tests/test_db_migrate.py [hunk 8]
```diff
@@ -72,16 +63,7 @@ def appliance_preupdate(temp_appliance_preconfig_funcscope_upgrade, appliance):
...
-@pytest.mark.tier(2)
-def test_db_migrate(temp_appliance_extended_db, db_url, db_version, db_desc):
-    """
-    Polarion:
-        assignee: jhenner
-        initialEstimate: 1/4h
-        casecomponent: Appliance
-    """
-    app = temp_appliance_extended_db
+def download_and_migrate_db(app, db_url, db_desc):
```
Comment (mshriver) [inline cfme/tests/test_db_migrate.py]: Optional / TODO

Consider moving this to utils? or defining on IPAppliance class as a method.

File: cfme/tests/services/test_service_performance.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    Bugzilla:
+        1688937
+    """
+
+    app = appliance_with_performance_db
+    assert 50000 == app.rest_api.collections.services.count
+
+    my_service = MyService(app)
+    # Timeit seems to accept callable as well as string of Python code on cPython.
```
Comment (mshriver) [inline cfme/tests/services/test_service_performance.py]: :+1:

**Pull request merged**